### PR TITLE
Add `f16` and `f128` const eval for binary and unary operationations

### LINKED ISSUE #124583: ICE: `Invalid float op Neg`
snippet:
````rust
fn main() {
    let _ = -(-0.0f16);
}


````

Version information
````
rustc 1.80.0-nightly (378a43a06 2024-05-01)
binary: rustc
commit-hash: 378a43a06510f3e3a49c69c8de71745e6a884048
commit-date: 2024-05-01
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: the type `f16` is unstable
 --> /tmp/icemaker_global_tempdir.LUs0yTDS2O3R/rustc_testrunner_tmpdir_reporting.laee65vXPGyd/mvce.rs:2:16
  |
2 |     let _ = -(-0.0f16);
  |                ^^^^^^
  |
  = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
  = help: add `#![feature(f16)]` to the crate attributes to enable
  = note: this compiler was built on 2024-05-01; consider upgrading it if it is out of date

error: internal compiler error: /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/compiler/rustc_const_eval/src/interpret/operator.rs:498:26: Invalid float op Neg
 --> /tmp/icemaker_global_tempdir.LUs0yTDS2O3R/rustc_testrunner_tmpdir_reporting.laee65vXPGyd/mvce.rs:1:1
  |
1 | fn main() {
  | ^^^^^^^^^

thread 'rustc' panicked at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/compiler/rustc_const_eval/src/interpret/operator.rs:498:26:
Box<dyn Any>
stack backtrace:
   0:     0x75fcab3ba035 - std::backtrace_rs::backtrace::libunwind::trace::hda47b82a166812ce
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x75fcab3ba035 - std::backtrace_rs::backtrace::trace_unsynchronized::hfdfc54f1e69a0c51
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x75fcab3ba035 - std::sys_common::backtrace::_print_fmt::h793b3f7571d13dfd
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x75fcab3ba035 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbc9f1fbbb950dc64
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x75fcab40929b - core::fmt::rt::Argument::fmt::h616c7e5e48e69035
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/core/src/fmt/rt.rs:165:63
   5:     0x75fcab40929b - core::fmt::write::h273cb7cce9dfa202
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/core/src/fmt/mod.rs:1157:21
   6:     0x75fcab3aebdf - std::io::Write::write_fmt::h945d30311b10e894
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/io/mod.rs:1832:15
   7:     0x75fcab3b9e0e - std::sys_common::backtrace::_print::hcdebe37872d00154
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x75fcab3b9e0e - std::sys_common::backtrace::print::h51bd45a216cc9df6
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x75fcab3bc779 - std::panicking::default_hook::{{closure}}::hb6bc4eb0d96a4bfd
  10:     0x75fcab3bc4bd - std::panicking::default_hook::h557ff9f57d861b48
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/panicking.rs:298:9
  11:     0x75fca7c3c6bb - std[8fa947ca3cb89c52]::panicking::update_hook::<alloc[4c35cc17dd12f569]::boxed::Box<rustc_driver_impl[f52ddb602ce7d3c0]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x75fcab3bce76 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hf5d9c19b8a775637
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/alloc/src/boxed.rs:2036:9
  13:     0x75fcab3bce76 - std::panicking::rust_panic_with_hook::h310a921a123719cf
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/panicking.rs:799:13
  14:     0x75fca7c6c084 - std[8fa947ca3cb89c52]::panicking::begin_panic::<rustc_errors[dbe3df606ebeeaa3]::ExplicitBug>::{closure#0}
  15:     0x75fca7c68d06 - std[8fa947ca3cb89c52]::sys_common::backtrace::__rust_end_short_backtrace::<std[8fa947ca3cb89c52]::panicking::begin_panic<rustc_errors[dbe3df606ebeeaa3]::ExplicitBug>::{closure#0}, !>
  16:     0x75fca7c689e6 - std[8fa947ca3cb89c52]::panicking::begin_panic::<rustc_errors[dbe3df606ebeeaa3]::ExplicitBug>
  17:     0x75fca7c75261 - <rustc_errors[dbe3df606ebeeaa3]::diagnostic::BugAbort as rustc_errors[dbe3df606ebeeaa3]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  18:     0x75fca822e208 - <rustc_errors[dbe3df606ebeeaa3]::DiagCtxt>::span_bug::<rustc_span[28b37812f336c7b5]::span_encoding::Span, alloc[4c35cc17dd12f569]::string::String>
  19:     0x75fca824f6bd - rustc_middle[ec0047e274fe25be]::util::bug::opt_span_bug_fmt::<rustc_span[28b37812f336c7b5]::span_encoding::Span>::{closure#0}
  20:     0x75fca824f6ea - rustc_middle[ec0047e274fe25be]::ty::context::tls::with_opt::<rustc_middle[ec0047e274fe25be]::util::bug::opt_span_bug_fmt<rustc_span[28b37812f336c7b5]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x75fca824874b - rustc_middle[ec0047e274fe25be]::ty::context::tls::with_context_opt::<rustc_middle[ec0047e274fe25be]::ty::context::tls::with_opt<rustc_middle[ec0047e274fe25be]::util::bug::opt_span_bug_fmt<rustc_span[28b37812f336c7b5]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x75fca8247f67 - rustc_middle[ec0047e274fe25be]::util::bug::span_bug_fmt::<rustc_span[28b37812f336c7b5]::span_encoding::Span>
  23:     0x75fcaa92bdef - <rustc_const_eval[c0dfeea2f3bf4919]::interpret::eval_context::InterpCx<rustc_const_eval[c0dfeea2f3bf4919]::const_eval::dummy_machine::DummyMachine>>::overflowing_unary_op.cold
  24:     0x75fca67bda02 - <rustc_mir_transform[61a8aa7cac63b4d6]::known_panics_lint::ConstPropagator as rustc_middle[ec0047e274fe25be]::mir::visit::Visitor>::visit_assign
  25:     0x75fca993d83d - <rustc_mir_transform[61a8aa7cac63b4d6]::known_panics_lint::ConstPropagator as rustc_middle[ec0047e274fe25be]::mir::visit::Visitor>::visit_body
  26:     0x75fca6999566 - <rustc_mir_transform[61a8aa7cac63b4d6]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[61a8aa7cac63b4d6]::pass_manager::MirLint>::run_lint
  27:     0x75fca90018cd - rustc_mir_transform[61a8aa7cac63b4d6]::pass_manager::run_passes_inner
  28:     0x75fca9395d42 - rustc_mir_transform[61a8aa7cac63b4d6]::mir_drops_elaborated_and_const_checked
  29:     0x75fca93955fd - rustc_query_impl[1888158ad4449b9b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[1888158ad4449b9b]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 8usize]>>
  30:     0x75fca938d171 - rustc_query_system[b5ee5054546adc93]::query::plumbing::try_execute_query::<rustc_query_impl[1888158ad4449b9b]::DynamicConfig<rustc_query_system[b5ee5054546adc93]::query::caches::VecCache<rustc_span[28b37812f336c7b5]::def_id::LocalDefId, rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[1888158ad4449b9b]::plumbing::QueryCtxt, false>
  31:     0x75fca938cb2a - rustc_query_impl[1888158ad4449b9b]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  32:     0x75fca936cff9 - rustc_interface[e7d426f81a3c67f]::passes::analysis
  33:     0x75fca936bddb - rustc_query_impl[1888158ad4449b9b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[1888158ad4449b9b]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 1usize]>>
  34:     0x75fca9dfb4e5 - rustc_query_system[b5ee5054546adc93]::query::plumbing::try_execute_query::<rustc_query_impl[1888158ad4449b9b]::DynamicConfig<rustc_query_system[b5ee5054546adc93]::query::caches::SingleCache<rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[1888158ad4449b9b]::plumbing::QueryCtxt, false>
  35:     0x75fca9dfb249 - rustc_query_impl[1888158ad4449b9b]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  36:     0x75fca9c675ce - rustc_interface[e7d426f81a3c67f]::interface::run_compiler::<core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>, rustc_driver_impl[f52ddb602ce7d3c0]::run_compiler::{closure#0}>::{closure#1}
  37:     0x75fca9c54167 - std[8fa947ca3cb89c52]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_with_globals<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_pool_with_globals<rustc_interface[e7d426f81a3c67f]::interface::run_compiler<core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>, rustc_driver_impl[f52ddb602ce7d3c0]::run_compiler::{closure#0}>::{closure#1}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>
  38:     0x75fca9c53f2a - <<std[8fa947ca3cb89c52]::thread::Builder>::spawn_unchecked_<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_with_globals<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_pool_with_globals<rustc_interface[e7d426f81a3c67f]::interface::run_compiler<core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>, rustc_driver_impl[f52ddb602ce7d3c0]::run_compiler::{closure#0}>::{closure#1}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#2} as core[19fe836d650da945]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x75fcab3c6cab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h32bd23449113524a
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/alloc/src/boxed.rs:2022:9
  40:     0x75fcab3c6cab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h04c486968e15e63c
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/alloc/src/boxed.rs:2022:9
  41:     0x75fcab3c6cab - std::sys::pal::unix::thread::Thread::new::thread_start::ha93fdc6af5a412fe
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys/pal/unix/thread.rs:108:17
  42:     0x75fcab16555a - <unknown>
  43:     0x75fcab1e2a3c - <unknown>
  44:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.80.0-nightly (378a43a06 2024-05-01) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_drops_elaborated_and_const_checked] elaborating drops for `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

### PR DESCRIPTION
**Pull request by tgross35**
Add const evaluation and Miri support for f16 and f128, including unary and binary operations. Casts are not yet included.

Fixes https://github.com/rust-lang/rust/issues/124583

r? @RalfJung

### CODE REVIEW

**Comment by tgross35** [comment]
src/tools/miri/tests/pass/float.rs is _really_ tedious to update and I'm not entirely clear what a lot of the magic numbers are intended to test. @RalfJung do you have any suggestions about how this could be cleaned up or made generic to some extent? Compiler-builtins uses a bit of fuzzing + property testing to compare the builtins function to the system function or apfloat https://github.com/rust-lang/compiler-builtins/blob/c04eb9e1afb72bdf943f5e5d77b3812f40526602/testcrate/src/lib.rs#L255, but we don't really have anything to compare it to here.

**Comment by RalfJung** [comment]
The file is kind of grown over time, and I did at some point copy a whole bunch of corner case tests over from the rustc test suite -- most of the cast tests are from that, I think.

But I don't know a good way to make this generic, no.

**Comment by tgross35** [comment]
That makes more sense if they came from the rust suite. I just removed the const eval casting until I can take a closer look at that and hopefully clean up that test a bit. Having only unary and binary operations still saves a ton of the headache that was blocking me elsewhere (ICE on negative numbers 😠)

Ready for a look when you get the chance.

File: tests/crashes/124583.rs [hunk 1]
```diff
@@ -1,5 +0,0 @@
-//@ known-bug: rust-lang/rust#124583
-
-fn main() {
-    let _ = -(-0.0f16);
-}
```
Comment (matthiaskrgr) [inline tests/crashes/124583.rs:5]: Could you add   Fixes #124583 to pr header so that this is auto closed? thx!

Comment (tgross35) [reply tests/crashes/124583.rs:5]: I had it in a commit message and thought it would snag it there, but added just in case :)

Comment (matthiaskrgr) [reply tests/crashes/124583.rs:5]: ah I didn't check the commits :upside_down_face: should be good enough then :)

**Comment by tgross35** [comment]
Spurious failure: https://github.com/rust-lang/rust/issues/126430

File: library/std/src/f128/tests.rs [hunk 2]
```diff
@@ -38,3 +38,9 @@ macro_rules! assert_f128_biteq {
         );
     };
 }
+
+#[test]
+fn test_neg_zero() {
+    // only verify const eval negation until we add the rest of the checks
+    let _neg_zero: f16 = -0.0;
+}
```
Comment (oli-obk) [inline library/std/src/f128/tests.rs:46]: These are thoroughly covered by the miri tests, I don't think we need it here, but if you wanna keep it, this needs to be f128

Comment (RalfJung) [reply library/std/src/f128/tests.rs:46]: Also this isn't even doing any const-eval? If anything it might be hitting some mir-opt codepaths.

Comment (oli-obk) [reply library/std/src/f128/tests.rs:46]: yea, const prop was hitting the mir interp code

Comment (tgross35) [reply library/std/src/f128/tests.rs]: Ah thanks for the catch, I missed that it was double negation that caused the crash in #124583. Updated `tests/ui/numbers-arithmetic/f16-f128-lit.rs` instead

File: src/tools/miri/tests/pass/float.rs [hunk 3]
```diff
@@ -41,104 +43,43 @@ trait FloatToInt<Int>: Copy {
...
+                }
+            }
+        )*
+    };
 }
 
+float_to_int_fallback!(f16 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int!(f32=> i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int!(f64 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int_fallback!(f128 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/float.rs:81]: The new impls aren't actually used, are they?

**Comment by RalfJung** [comment]
The last commit seems to un-do a bunch of the changes from the earlier commits...?

**Comment by tgross35** [comment]
> The last commit seems to un-do a bunch of the changes from the earlier commits...?

I meant to remove the changes to cast entirely since I don't have a good test (getting tedious as mentioned above), just did a bad job rebasing. History should be cleaner now.

File: src/tools/miri/src/helpers.rs [hunk 4]
```diff
@@ -1201,12 +1201,14 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
-            FloatTy::F16 => unimplemented!("f16_f128"),
+            FloatTy::F16 =>
+                float_to_int_inner::<Half>(this, src.to_scalar().to_f16()?, cast_to, round),
             FloatTy::F32 =>
                 float_to_int_inner::<Single>(this, src.to_scalar().to_f32()?, cast_to, round),
             FloatTy::F64 =>
                 float_to_int_inner::<Double>(this, src.to_scalar().to_f64()?, cast_to, round),
-            FloatTy::F128 => unimplemented!("f16_f128"),
+            FloatTy::F128 =>
+                float_to_int_inner::<Quad>(this, src.to_scalar().to_f128()?, cast_to, round),
```
Comment (RalfJung) [inline src/tools/miri/src/helpers.rs:1211]: This seems to be dead code now that you removed the cast tests?

OTOH, the code is rather obvious, so I am also fine with keeping it.

**Comment by RalfJung** [comment]
Maybe update the PR title and description to say what is covered? Mainly unops and binops, it seems.

r=me after that, with or without the helpers.rs change.

File: src/tools/miri/src/helpers.rs (hunk 4 above)
Comment (tgross35) [reply src/tools/miri/src/helpers.rs:1211]: Yeah you are right. If you have no problem then I think I may as well just leave it, casts won't be far behind.

**Comment by tgross35** [comment]
I updated the description, feel free to tweak it if I'm missing something. (You or someone else need to r for me).

@rustbot label +F-f16_and_f128

**Pull request merged**